File: internal/actions/mailer.go
package actions

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"log/slog"
	"mime"
	"net"
	"net/mail"
	"net/smtp"
	"strconv"
	"strings"
	"time"

	"github.com/openforms/openforms/internal/config"
)

type Mailer interface {
	Send(ctx context.Context, to, subject, body string) error
}

// NewMailer returns an SMTP mailer when SMTPHost is configured, otherwise a
// mailer that only logs (useful for local development).
func NewMailer(cfg config.Config) Mailer {
	if cfg.SMTPHost == "" {
		return LogMailer{}
	}
	return &SMTPMailer{
		Addr:     net.JoinHostPort(cfg.SMTPHost, strconv.Itoa(cfg.SMTPPort)),
		Host:     cfg.SMTPHost,
		Username: cfg.SMTPUsername,
		Password: cfg.SMTPPassword,
		From:     cfg.SMTPFrom,
	}
}

type LogMailer struct{}

func (LogMailer) Send(ctx context.Context, to, subject, body string) error {
	slog.InfoContext(ctx, "email (log mailer)", "to", to, "subject", subject, "body", body)
	return nil
}

type SMTPMailer struct {
	Addr, Host, Username, Password, From string
	// SendFunc defaults to smtp.SendMail; tests replace it.
	SendFunc func(addr string, a smtp.Auth, from string, to []string, msg []byte) error
}

func (m *SMTPMailer) Send(ctx context.Context, to, subject, body string) error {
	msg, err := BuildMessage(m.From, to, subject, body, time.Now())
	if err != nil {
		return err
	}
	var auth smtp.Auth
	if m.Username != "" {
		auth = smtp.PlainAuth("", m.Username, m.Password, m.Host)
	}
	send := m.SendFunc
	if send == nil {
		send = smtp.SendMail
	}
	if err := send(m.Addr, auth, m.From, []string{to}, msg); err != nil {
		return fmt.Errorf("smtp send: %w", err)
	}
	return nil
}

// BuildMessage renders a plain-text RFC 5322 message. It rejects addresses
// containing CR/LF and strips CR/LF from the subject to prevent header injection.
func BuildMessage(from, to, subject, body string, date time.Time) ([]byte, error) {
	for _, a := range []string{from, to} {
		if strings.ContainsAny(a, "\r\n") {
			return nil, errors.New("email address contains a line break")
		}
	}
	if _, err := mail.ParseAddress(to); err != nil {
		return nil, fmt.Errorf("invalid recipient %q: %w", to, err)
	}
	subject = strings.NewReplacer("\r\n", " ", "\r", " ", "\n", " ").Replace(subject)
	body = strings.ReplaceAll(strings.ReplaceAll(body, "\r\n", "\n"), "\n", "\r\n")

	var b bytes.Buffer
	fmt.Fprintf(&b, "From: %s\r\n", from)
	fmt.Fprintf(&b, "To: %s\r\n", to)
	fmt.Fprintf(&b, "Subject: %s\r\n", mime.QEncoding.Encode("utf-8", subject))
	fmt.Fprintf(&b, "Date: %s\r\n", date.Format(time.RFC1123Z))
	b.WriteString("MIME-Version: 1.0\r\n")
	b.WriteString("Content-Type: text/plain; charset=utf-8\r\n")
	b.WriteString("Content-Transfer-Encoding: 8bit\r\n")
	b.WriteString("\r\n")
	b.WriteString(body)
	return b.Bytes(), nil
}

File: src/openforms/server/actions/mailer.py
"""SMTP and log mailers; ``build_message`` renders a plain-text RFC 5322 message."""

from __future__ import annotations

import datetime as dt
import logging
import smtplib
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Protocol

import anyio

from ...definition.common import is_bare_email
from ...settings import Settings

log = logging.getLogger("openforms.mail")


class Mailer(Protocol):
    async def send(self, to: str, subject: str, body: str) -> None: ...


class LogMailer:
    """Logs instead of sending (no SMTP host configured)."""

    async def send(self, to: str, subject: str, body: str) -> None:
        log.info("email (log mailer) to=%s subject=%s body=%s", to, subject, body)


SendFunc = Callable[[str, int, "tuple[str, str] | None", str, list[str], bytes], None]


def _smtp_send(host: str, port: int, auth: tuple[str, str] | None, from_: str, to: list[str], msg: bytes) -> None:
    with smtplib.SMTP(host, port, timeout=30) as s:
        s.ehlo()
        if s.has_extn("starttls"):
            s.starttls()
            s.ehlo()
        if auth is not None:
            s.login(*auth)
        s.sendmail(from_, to, msg)


@dataclass
class SMTPMailer:
    host: str
    port: int
    from_: str
    username: str = ""
    password: str = ""
    send_func: SendFunc = field(default=_smtp_send)

    @property
    def addr(self) -> str:
        return f"{self.host}:{self.port}"

    async def send(self, to: str, subject: str, body: str) -> None:
        msg = build_message(self.from_, to, subject, body, dt.datetime.now(dt.UTC))
        auth = (self.username, self.password) if self.username else None
        try:
            await anyio.to_thread.run_sync(self.send_func, self.host, self.port, auth, self.from_, [to], msg)
        except Exception as e:
            raise RuntimeError(f"smtp send: {e}") from e


def new_mailer(settings: Settings) -> Mailer:
    if not settings.smtp_host:
        return LogMailer()
    return SMTPMailer(
        host=settings.smtp_host,
        port=settings.smtp_port,
        from_=settings.smtp_from,
        username=settings.smtp_username,
        password=settings.smtp_password,
    )


def _needs_encoding(s: str) -> bool:
    return any(not (" " <= c <= "~" or c == "\t") for c in s)


def q_encode(s: str) -> str:
    """RFC 2047 Q-encoding as Go's ``mime.QEncoding.Encode("utf-8", s)`` does it."""
    if not _needs_encoding(s):
        return s
    prefix, suffix, max_word = "=?utf-8?q?", "?=", 75
    words: list[str] = []
    cur = ""
    for ch in s:
        enc = ""
        for b in ch.encode("utf-8"):
            if b == 0x20:
                enc += "_"
            elif 0x21 <= b <= 0x7E and b not in (ord("="), ord("?"), ord("_")):
                enc += chr(b)
            else:
                enc += f"={b:02X}"
        if len(prefix) + len(cur) + len(enc) + len(suffix) > max_word:
            words.append(prefix + cur + suffix)
            cur = ""
        cur += enc
    words.append(prefix + cur + suffix)
    return " ".join(words)


def build_message(from_: str, to: str, subject: str, body: str, date: dt.datetime) -> bytes:
    """Reject addresses containing CR/LF and strip CR/LF from the subject (no header injection)."""
    for a in (from_, to):
        if "\r" in a or "\n" in a:
            raise ValueError("email address contains a line break")
    if not is_bare_email(to):
        raise ValueError(f"invalid recipient {to!r}")
    subject = subject.replace("\r\n", " ").replace("\r", " ").replace("\n", " ")
    body = body.replace("\r\n", "\n").replace("\n", "\r\n")
    lines = [
        f"From: {from_}",
        f"To: {to}",
        f"Subject: {q_encode(subject)}",
        f"Date: {date.strftime('%a, %d %b %Y %H:%M:%S %z')}",
        "MIME-Version: 1.0",
        "Content-Type: text/plain; charset=utf-8",
        "Content-Transfer-Encoding: 8bit",
        "",
    ]
    return ("\r\n".join(lines) + "\r\n" + body).encode("utf-8")
